File: Algorithms & Data Structures/Data structures implementation JS/Queue.js
class Node {
    constructor(item) {
        this.item = item;
        this.right = null; // alias for 'next'
        this.front = null;
        this.rear = null;
    }
}

class Queue {
    constructor() {
        this.length = 0;
        this.front = null;
        this.rear = null;
    }
    isEmpty() {
        return this.front == null && this.rear == null;
    }
    enqueue(item) {
        let newNode = new Node(item);
        if (this.isEmpty()) {
            this.front = newNode;
            this.rear = newNode;
        }
        else {
            this.rear.right = newNode;
            this.rear = newNode;
        }
        this.length++;
    }
    dequeue() {
        if (!this.isEmpty()) {
            if (this.front == this.rear) {
                this.front == null;
                this.rear = null;
            }
            else {
                this.front = this.front.right;
            }
            this.length--;
        }
        else {
            console.log("Queue is empty!");
        }
    }
    print() {
        let current = this.front;
        while (current) {
            console.log(current.item + "\n");
            current = current.right;
        }
    }
}

File: Algorithms & Data Structures/Data structures implementation C++/Queue.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cstdlib>
using namespace std;

struct node {
  int data;
  node *next;
  node *parent;
  node *left;
  node *right;
};


// Queue //

class queue {
  public:
    node *front, *rear;
    queue() {
      front = NULL;
      rear = NULL;
    }
    void enqueue(int item) {
      node *temp = new node;
      temp->data = item;
      temp->next = NULL;
      if(front == NULL) {
        front = temp;
        rear = temp;
      }
      else {
        rear->next = temp;
        rear = temp;
      }
    }
    int dequeue() {
      node *old = front;
      front = front->next;
      if(front == NULL) {
        rear = NULL; // I'm not sure of importance of this line :D //
      }
      int oldData = old->data;
      delete old;
      return oldData;
    }
};

// Queue END //

int main() {
 // Queue: enqueue, dequeue
 queue q;
}